src/Random.mo

```diff
@@ -8,63 +18,206 @@ module {
+          let m = mask;
+          mask >>= (1 : Nat8);
+          ?(0 : Nat8 != byte & m)
+        }
+      }
+    };
 
     /// Random choice between `true` and `false`.
     public func bool() : Bool {
-      todo()
+      switch (bitIter.next()) {
+        case (?bit) { bit };
+        case null {
+          iter := generator().vals();
+          switch (bitIter.next()) {
+            case (?bit) { bit };
+            case null {
+              Runtime.trap("Random.bool(): generator produced empty Blob")
+            }
+          }
+        }
+      }
     };
 
     /// Random `Nat8` value in the range [0, 256).
-    public func byte() : Nat8 {
-      todo()
+    public func nat8() : Nat8 {
+      switch (iter.next()) {
+        case (?byte) { byte };
+        case null {
+          iter := generator().vals();
+          switch (iter.next()) {
+            case (?byte) { byte };
+            case null {
+              Runtime.trap("Random.nat8(): generator produced empty Blob")
+            }
+          }
+        }
+      }
     };
 
-    /// Random `Float` value in the range [0, 1).
-    /// The resolution is maximal (i.e. all relevant mantissa bits are randomized).
-    public func float() : Float {
-      todo()
+    // Helper function which returns a uniformly sampled `Nat64` in the range `[0, toExclusive)`.
+    // Uses rejection sampling to ensure uniform distribution even when the range
+    // doesn't divide evenly into 2^64. This avoids modulo bias that would occur
+    // from simply taking the modulo of a random 64-bit number.
+    func uniform64(toExclusive : Nat64) : Nat64 {
+      if (toExclusive <= 1) {
+        return 0
+      };
+      // 2^64 - (2^64 % toExclusive) = (2^64-1) - (2^64-1 % toExclusive):
+      let cutOff = Nat64.maxValue - (Nat64.maxValue % toExclusive);
+      // 2^64 / toExclusive, with toExclusive > 1:
+      let multiple = Nat64.fromNat(Nat.pow(2, 64) / Nat64.toNat(toExclusive));
+      loop {
+        // Build up a random Nat64 from bytes
+        var number : Nat64 = 0;
+        for (_ in Nat8.range(0, 8)) {
+          number := (number << 8) | Nat64.fromNat(Nat8.toNat(nat8()))
+        };
+        // If number is below cutoff, we can use it
+        if (number < cutOff) {
+          // Scale down to desired range
+          return number / multiple
+        };
+        // Otherwise reject and try again
+      }
     };
 
-    public func intRange(min : Int, maxExclusive : Int) : Int {
-      todo()
+    public func nat64() : Nat64 {
+      uniform64(Nat64.maxValue)
     };
 
-    public func natRange(min : Nat, maxExclusive : Nat) : Nat {
-      todo()
+    public func nat64Range(from : Nat64, toExclusive : Nat64) : Nat64 {
+      if (from > toExclusive) {
+        Runtime.trap("Random.nat64Range(): from > toExclusive")
+      };
+      uniform64(toExclusive - from) + from
+    };
+
+    public func intRange(from : Int, toExclusive : Int) : Int {
+      switch (Nat.fromInt(toExclusive - from)) {
```

Comment:
Does Nat.fromInt return an option? That would be different from all the other conversions that trap.... I wonder if we should revisit that choice....